File: Crash it/StarsGenerator.swift
//
//  Stars.swift
//  Crash it
//
//  Created by Olivier Picard on 15/03/2018.
//  Copyright © 2018 Olivier Picard. All rights reserved.
//

import Foundation
import SpriteKit

/**
 Gestion semi-automatisée de l'ensemble des étoiles sur l'écran
 */
class StarsGenerator {
    private var stars_percent: Int
    private let screen_size: CGSize
    private var current_scene: SKScene
    var star_size: CGSize
    var speed_factor: Float
    var opacity_range: Tools.Interval
    
    init(scene: SKScene, screenSize: CGSize, starsPercent: Int,
         opacityRange: Tools.Interval, starSize: CGSize, speedFactor: Float) {
        self.screen_size = screenSize
        self.current_scene = scene
        self.stars_percent = starsPercent
        self.speed_factor = speedFactor
        self.opacity_range = opacityRange
        self.star_size = starSize
        
        // Crée aléatoirement des étoiles sur l'écran
        // Pour ne pas commencer le jeu avec un écran vide
        for _ in 0...stars_percent*10 {
            let randHeight = arc4random_uniform(UInt32(self.screen_size.height))
            let randWidth = arc4random_uniform(UInt32(self.screen_size.width))
            let star = MovingItem(texture: nil, size: self.star_size)
            star.speed_factor = self.speed_factor
            star.position = CGPoint(x: Int(randWidth), y: Int(randHeight))
            star.color = star.color.withAlphaComponent(CGFloat(opacity_range.random()) / 100)
            star.zPosition = 10
            self.current_scene.addChild(star)
        }
    }
    
    
    /**
     Génère une étoile au sommet de l'écran (en dehors de la zone visible)
     et aléatoirement sur l'axe X
     La pourcentage d'apparation est défini
     à l'initialisation par 'percentageOfStars'
    */
    func generate() {
        // Si le nombre aléatoire n'est pas compris dans [0, percentage_of_stars]
        // alors on arrête, ne gérère pas l'étoile (pour le prochain
        // update si la chance est de notre côté)
        if(arc4random_uniform(101) > stars_percent*Int(MovingItem.base_moving_speed)) { return }
        
        // hauteur de l'écran + 10 pour être sûre que
        // l'étoile sera générer en dehors de l'écran
        let yPos = screen_size.height + 10
        let xPos = CGFloat(arc4random_uniform(UInt32(screen_size.width)))
        
        let star = MovingItem(texture: nil, size: star_size)
        star.speed_factor = self.speed_factor
        star.position = CGPoint(x: xPos, y: yPos)
        star.color = UIColor.white
        star.alpha = CGFloat(opacity_range.random()) / 100
        star.zPosition = 10
        self.current_scene.addChild(star)
        
    }
    
    
    
    
    
}

File: Android/app/src/main/java/com/olivierpicard/crachit/StarsGenerator.java
package com.olivierpicard.crachit;

import android.graphics.Color;

import com.olivierpicard.crachit.Graphics.GInterval;
import com.olivierpicard.crachit.Graphics.GPoint;
import com.olivierpicard.crachit.Graphics.GScene;
import com.olivierpicard.crachit.Graphics.GSize;
import com.olivierpicard.crachit.Graphics.GVector;

import java.util.concurrent.ThreadLocalRandom;

/**
 * Created by olivierpicard on 04/04/2018.
 */

public class StarsGenerator {
    private int stars_percent;
    private GameScene current_scene;
    GSize star_size;
    float speed_factor;
    GInterval opacity_range;

    StarsGenerator(GameScene scene,
                   int starsPercent,
                   GInterval opacityRange,
                   GSize starSize,
                   float speedFactor)
    {
        this.current_scene = scene;
        this.stars_percent = starsPercent;
        this.speed_factor = speedFactor;
        this.opacity_range = opacityRange;
        this.star_size = starSize;

        // Crée aléatoirement des étoiles sur l'écran
        // Pour ne pas commencer le jeu avec un écran vide
        for(int i = 0; i < starsPercent*10; i++) {
            final int randHeight = GInterval.random(0, Tools.screenMetrics.heightPixels);
            final int randWidth = GInterval.random(0, Tools.screenMetrics.widthPixels);
            MovingItem star = new MovingItem(null, Color.WHITE, this.star_size);
            star.speed_factor = this.speed_factor;
            star.setPosition(new GPoint(randWidth, randHeight));
            star.setZPosition(10);
            star.setColor(Tools.setColorOpacity(Color.WHITE, opacity_range.random()));
            this.current_scene.addChild(star);
        }
    }


    void generate() {
        // Si le nombre aléatoire n'est pas compris dans [0, percentage_of_stars]
        // alors on arrête, ne gérère pas l'étoile (pour le prochain
        // update si la chance est de notre côté)
        if(GInterval.random(0, 100) > stars_percent*MovingItem.base_moving_speed) return;

        // hauteur de l'écran + 10 pour être sûre que
        // l'étoile sera générer en dehors de l'écran
        int yPos = -10;
        int xPos = GInterval.random(0, Tools.screenMetrics.widthPixels);

        MovingItem star = new MovingItem(null, Color.WHITE, this.star_size);
        star.speed_factor = this.speed_factor;
        star.setPosition(new GPoint(xPos, yPos));
        star.setColor(Tools.setColorOpacity(Color.WHITE, opacity_range.random()));
        star.setZPosition(10);
        this.current_scene.addChild(star);
    }
}
